driver: espi: add espi driver for rts5912

JhanBoChao-Realtek opened the pull request:
add espi driver for rts5912.

albertofloyd commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -0,0 +1,2338 @@
+			espi_reg->ESPICFG = data->config_data;
+
+			if (espi_reg->EVCFG & ESPI_EPCFG_CHEN_Msk) {
+				k_work_schedule(&vw_ch_isr_manual_trigger, K_MSEC(150));
```
Avoid magic numbers, instead define and document a macro

JhanBoChao-Realtek replied:
done

albertofloyd commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -0,0 +1,2338 @@
+	uint32_t config = espi_reg->EPCFG;
+
+	if (status & ESPI_EPSTS_CLRSTS_Msk) {
+		evt.evt_data = (config & ESPI_EPCFG_CHEN_Msk) ? 1 : 0;
```
Ensure the data follows the polarity assumptions described on espi header

JhanBoChao-Realtek replied:
done

albertofloyd commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -0,0 +1,2338 @@
+	[signal] = {.vw_index = index, .level_mask = level, .valid_mask = valid}
+
+/* VW signals used in eSPI */
+static const struct espi_vw_channel vw_channel_list[] = {
```
Try to avoid magic numbers usage, either use HAL or macro definitions to ease mantainance

JhanBoChao-Realtek replied:
OK, thank you

albertofloyd commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -0,0 +1,2338 @@
+				}
+			}
+		}
+		if (espi_vw_ch_cached_data.idx2 == espi_reg->EVIDX2) {
```
nit, extra line after conditional block

albertofloyd commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -0,0 +1,2338 @@
+	}
+
+	switch (vw_index) {
+	case 0x04:
```
magic numbers

JhanBoChao-Realtek replied:
done

albertofloyd commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -0,0 +1,2338 @@
+
+	uint8_t vw_idx = vw_channel_list[signal].vw_index;
+	uint8_t lev_msk = vw_channel_list[signal].level_mask;
+	uint8_t rd_cnt = 100;
```
magic number

JhanBoChao-Realtek replied:
done

albertofloyd commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -0,0 +1,2338 @@
+
+	while (espi_reg->EVSTS & ESPI_EVSTS_TXFULL_Msk) {
+		timeout_cnt++;
+		if (timeout_cnt >= 10000) {
```
Add a macro to replace this magic number and document logic to poll for that amount of time

JhanBoChao-Realtek replied:
done

keith-zephyr commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -0,0 +1,2338 @@
+	.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(0),
+};
+
+DEVICE_DT_INST_DEFINE(0, &espi_rts5912_init, NULL, &espi_rts5912_data, &espi_rts5912_config,
```
Add a BUILD_ASSERT that checks that only one driver instance has been enabled.

JhanBoChao-Realtek replied:
done

JhanBoChao-Realtek committed:
driver: espi: add espi driver for rts5912

add espi driver for rts5912

Signed-off-by: jhan bo chao <jhan_bo_chao@realtek.com>

JhanBoChao-Realtek committed:
driver: espi: add espi flash channel driver for rts5912

add espi flash channel driver for rts5912

Signed-off-by: jhan bo chao <jhan_bo_chao@realtek.com>

JhanBoChao-Realtek committed:
driver: espi: add espi oob channel driver for rts5912

add espi oob channel driver for rts5912

Signed-off-by: jhan bo chao <jhan_bo_chao@realtek.com>

JhanBoChao-Realtek committed:
driver: espi: add espi vw channel driver for rts5912

add espi vw channel driver for rts5912

Signed-off-by: jhan bo chao <jhan_bo_chao@realtek.com>

JhanBoChao-Realtek committed:
driver: espi: add espi peripheral channel port 80 driver for rts5912

add espi peripheral channel port 80 driver for rts5912

Signed-off-by: jhan bo chao <jhan_bo_chao@realtek.com>

JhanBoChao-Realtek committed:
driver: espi: add espi peripheral channel HOST_CMD driver for rts5912

espi: add espi peripheral channel HOST_CMD driver for rts5912

Unlike other chips using IO port 0x800-0x8ff, we utilize shared memory to
transfer host command parameters. The AP firmware must have corresponding
settings for this configuration.

Signed-off-by: jhan bo chao <jhan_bo_chao@realtek.com>

JhanBoChao-Realtek committed:
driver: espi: add espi peripheral channel acpi driver for rts5912

add espi peripheral channel acpi driver for rts5912

Signed-off-by: jhan bo chao <jhan_bo_chao@realtek.com>

JhanBoChao-Realtek committed:
driver: espi: add espi peripheral channel acpi shd mem driver for rts5912

add espi peripheral channel acpi shd mem driver for rts5912

Signed-off-by: jhan bo chao <jhan_bo_chao@realtek.com>

JhanBoChao-Realtek committed:
driver: espi: add espi peripheral channel 8042_KBC driver for rts5912

add espi peripheral channel 8042_KBC driver for rts5912

Signed-off-by: jhan bo chao <jhan_bo_chao@realtek.com>

JhanBoChao-Realtek committed:
driver: espi: add espi peripheral channel driver for rts5912

add espi peripheral channel driver for rts5912

Signed-off-by: jhan bo chao <jhan_bo_chao@realtek.com>

JhanBoChao-Realtek commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -0,0 +1,2368 @@
+	uint32_t port80_base;
+	uint32_t port80_clk_grp;
+	uint32_t port80_clk_idx;
+	uint32_t acpi_base;
```
done

JhanBoChao-Realtek commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -0,0 +1,2368 @@
+	uint32_t espislv_base;
+	uint32_t espislv_clk_grp;
+	uint32_t espislv_clk_idx;
+	uint32_t port80_base;
```
done

JhanBoChao-Realtek commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -0,0 +1,2368 @@
+#define RTS5912_ESPI_RST_IRQ DT_INST_IRQ_BY_IDX(0, 0, irq)
+
+struct espi_rts5912_config {
+	uint32_t espislv_base;
```
done

JhanBoChao-Realtek commented on soc/realtek/ec/rts5912/reg/reg_emi.h:
```diff
@@ -0,0 +1,18 @@
+#ifndef ZEPHYR_SOC_REALTEK_RTS5912_REG_EMI_H
+#define ZEPHYR_SOC_REALTEK_RTS5912_REG_EMI_H
+
+typedef struct {
```
done

JhanBoChao-Realtek commented on drivers/espi/Kconfig.rts5912:
```diff
@@ -0,0 +1,45 @@
+		default y
+
+	config ESPI_PERIPHERAL_EC_HOST_CMD
+		default y
```
done

JhanBoChao-Realtek commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -0,0 +1,2338 @@
+
+static const struct espi_rts5912_config espi_rts5912_config = {
+	.espislv_base = DT_INST_REG_ADDR_BY_IDX(0, 0),
+	.espislv_clk_grp = DT_CLOCKS_CELL_BY_NAME(DT_NODELABEL(espi0), espi_slave, clk_grp),
```
done

keith-zephyr replied:
This still isn't fixed.  You're still using the nodelabel instead of the DT_DRV_INST

JhanBoChao-Realtek commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -0,0 +1,2338 @@
+static struct espi_rts5912_data espi_rts5912_data;
+
+static const struct espi_rts5912_config espi_rts5912_config = {
+	.espislv_base = DT_INST_REG_ADDR_BY_IDX(0, 0),
```
done

JhanBoChao-Realtek commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -0,0 +1,2338 @@
+	}
+
+	switch (vw_idx) {
+	case 0x04:
```
done

keith-zephyr commented on boards/realtek/rts5912_evb/rts5912_evb_defconfig:
```diff
@@ -23,3 +23,6 @@ CONFIG_INPUT=y
 CONFIG_ADC=y
+
+# Enable ESPI
+CONFIG_ESPI=y
```
This should be enabled in the [espi sample](https://github.com/zephyrproject-rtos/zephyr/blob/c0a423426f43a31eaf0b821c8433fc110f987bf1/samples/drivers/espi/README.rst#L4), not in the common board code.

JhanBoChao-Realtek replied:
done

keith-zephyr commented on drivers/espi/Kconfig.rts5912:
```diff
@@ -0,0 +1,10 @@
+	bool "Realtek RTS5912 eSPI Driver"
+	default y
+	depends on DT_HAS_REALTEK_RTS5912_ESPI_ENABLED
+	select KERNEL_MEM_POOL
```
`KERNEL_MEM_POOL` defaults to 'y'.  Why is this here?

JhanBoChao-Realtek replied:
done, removed.

keith-zephyr commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -0,0 +1,236 @@
+static int espi_rts5912_manage_callback(const struct device *dev, struct espi_callback *callback,
+					bool set)
+{
+	struct espi_rts5912_data *data = (struct espi_rts5912_data *)dev->data;
```
Cast isn't needed here or in any of the other routines.
```suggestion
	struct espi_rts5912_data *data = dev->data;
```

keith-zephyr commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -0,0 +1,236 @@
+			/* rst pin low go high trigger interrupt */
+			evt.evt_data = 1;
+		}
+		espi_send_callbacks(&data->callbacks, dev, evt);
```
It looks like this shouldn't compile?
```suggestion
		espi_send_callbacks(&data->callbacks, dev, &evt);
```

JhanBoChao-Realtek replied:
static inline void espi_send_callbacks(sys_slist_t *list,
				       const struct device *dev,
				       struct espi_event evt)
The function is pass evt by value

keith-zephyr replied:
My mistake - you're correct.

keith-zephyr commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -0,0 +1,236 @@
+	.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(0),
+};
+
+DEVICE_DT_INST_DEFINE(0, &espi_rts5912_init, NULL, &espi_rts5912_data_0, &espi_rts5912_config,
```
You need a BUILD_ASSERT() to check that only one `realtek,rts5912-espi` instance is defined.

JhanBoChao-Realtek replied:
done

keith-zephyr commented on dts/arm/realtek/ec/rts5912.dtsi:
```diff
@@ -167,6 +167,80 @@
+			"kbc";
+
+			interrupts = <133 0>, <134 0>, <146 0>,
+						<145 0>, <144 0>, <143 0>,
```
Odd indentation.  Please only indent one additional tab here and for the `interrupt-names` properties.

JhanBoChao-Realtek replied:
done

keith-zephyr commented on soc/realtek/ec/rts5912/reg/reg_espi.h:
```diff
@@ -0,0 +1,633 @@
+#define ZEPHYR_SOC_REALTEK_RTS5912_REG_ESPI_H
+
+struct espi_reg {
+	volatile uint32_t EPSTS;
```
Don't use `volatile` on the individual fields - the users of `struct espi_reg` should add the `volatile` keyword.

JhanBoChao-Realtek replied:
done, removed.

keith-zephyr commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -30,8 +30,217 @@ struct espi_rts5912_config {
+static int espi_rts5912_flash_read(const struct device *dev, struct espi_flash_packet *pckt)
+{
+	const struct espi_rts5912_config *const espi_config = dev->config;
+	struct espi_rts5912_data *espi_data = (struct espi_rts5912_data *)dev->data;
```
Drop the unneeded pointer cast.

JhanBoChao-Realtek replied:
done

keith-zephyr commented on soc/realtek/ec/rts5912/Kconfig.defconfig.rts5912:
```diff
@@ -22,4 +22,8 @@ config ARCH_HAS_CUSTOM_BUSY_WAIT
 
 endif # REALTEK_RTS5912_RTMR
 
+config HEAP_MEM_POOL_SIZE
+	default 8192
+	depends on ESPI_RTS5912
```
Why do you need malloc?

JhanBoChao-Realtek replied:
We use the "k_aligned_alloc" to allocate buffer in espi_flash_ch_setup and espi_oob_ch_setup.

keith-zephyr replied:
If you never release this memory, I prefer that you statically allocate the buffer instead of bringing in the malloc code.

But the driver shouldn't set the heap size directly, you should create a symbol with the prefix `HEAP_MEM_POOL_ADD_SIZE_` to reserve HEAP for the eSPI driver.  The board code needs to set `CONFIG_HEAP_MEM_POOL_SIZE`.  

```
config HEAP_MEM_POOL_ADD_SIZE_ESPI_RTS5912
    int
    default 8192 if ESPI_RTS5912
    default 0
```

keith-zephyr commented on drivers/espi/Kconfig.rts5912:
```diff
@@ -8,3 +8,10 @@ config ESPI_RTS5912
+
+if ESPI_RTS5912
+
+	config ESPI_FLASH_CHANNEL
+		default y
```
Don't indent inside `if/endif` Kconfig blocks.

JhanBoChao-Realtek replied:
done

keith-zephyr commented on drivers/espi/Kconfig.rts5912:
```diff
@@ -11,6 +11,9 @@ config ESPI_RTS5912
 
 if ESPI_RTS5912
 
+	config ESPI_OOB_CHANNEL
+		default y
```
Don't indent this.

JhanBoChao-Realtek replied:
done

keith-zephyr commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -30,12 +30,231 @@ struct espi_rts5912_config {
+static int espi_rts5912_send_oob(const struct device *dev, struct espi_oob_packet *pckt)
+{
+	const struct espi_rts5912_config *const espi_config = dev->config;
+	struct espi_rts5912_data *espi_data = (struct espi_rts5912_data *)dev->data;
```
Delete unneeded pointer cast.

JhanBoChao-Realtek replied:
done

keith-zephyr commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -30,12 +30,231 @@ struct espi_rts5912_config {
+
+	volatile struct espi_reg *const espi_reg = espi_config->espi_reg;
+
+#ifndef CONFIG_ESPI_OOB_CHANNEL_RX_ASYNC
+	int ret;
+#endif
```
You can declare `ret` inside the #ifndef block below.

JhanBoChao-Realtek replied:
done

keith-zephyr replied:
You didn't make this change.

JhanBoChao-Realtek replied:
done, sorry.

keith-zephyr commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -30,12 +30,231 @@ struct espi_rts5912_config {
+
+#ifndef CONFIG_ESPI_OOB_CHANNEL_RX_ASYNC
+	/* Wait until ISR or timeout */
+	ret = k_sem_take(&espi_data->oob_rx_lock, K_MSEC(MAX_OOB_TIMEOUT));
```
```suggestion
	int ret = k_sem_take(&espi_data->oob_rx_lock, K_MSEC(MAX_OOB_TIMEOUT));
```

JhanBoChao-Realtek replied:
done

keith-zephyr replied:
You didn't make this change.

JhanBoChao-Realtek replied:
done, sorry.

keith-zephyr commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -45,6 +45,895 @@ struct espi_rts5912_data {
+static K_WORK_DELAYABLE_DEFINE(vw_ch_isr_wa, vw_ch_isr_wa_cb);
+
+#ifdef CONFIG_ESPI_AUTOMATIC_BOOT_DONE_ACKNOWLEDGE
+static void send_slave_bootdone(const struct device *dev)
```
Switch to the inclusive language terms `controller` and `target`.  `send_target_bootdone`


https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-2-inclusive-language

JhanBoChao-Realtek replied:
done.

keith-zephyr commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -1242,6 +1468,92 @@ static void espi_vw_ch_setup(const struct device *dev)
+		return -EBUSY;
+	}
+
+	while (espi_reg->EVSTS & ESPI_EVSTS_TXFULL) {
+		timeout_cnt++;
+		if (timeout_cnt >= ESPI_VW_EVENT_TIMEOUT_CNT) {
+			return -EBUSY;
+		}
+	}
```
Non-blocking comment.  You can use the `WAIT_FOR()` macro to make this easier to read.
```
if (!WAIT_FOR((espi_reg->EVSTS & ESPI_EVSTS_TXFULL) == 0, <time in microseconds>, k_busy_wait(10))) {
    return -EBUSY;
}
```

JhanBoChao-Realtek replied:
done

keith-zephyr commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -1242,6 +1468,92 @@ static void espi_vw_ch_setup(const struct device *dev)
+	i = 0x0000FFFF & ((uint32_t)index << 8 | (uint32_t)data);
+	espi_reg->EVTXDAT = i;
+
+	while (espi_reg->EVSTS & ESPI_EVSTS_TXFULL && retry_counter) {
+		retry_counter--;
+	}
```
Non-blocking - this can also use `WAIT_FOR` to make this timeout predictable.

JhanBoChao-Realtek replied:
done

keith-zephyr commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -1242,6 +1468,92 @@ static void espi_vw_ch_setup(const struct device *dev)
+		return;
+	}
+
+	while (espi_reg->EVSTS & ESPI_EVSTS_TXFULL) {
+		timeout_cnt++;
+		if (timeout_cnt >= ESPI_VW_EVENT_TIMEOUT_CNT) {
+			return;
+		}
+	}
```
Non-blocking: another place to use `WAIT_FOR()`

You also check the same condition `TXFULL` with the same timeout in multiple places.  This can be moved to a helper function.

JhanBoChao-Realtek commented on drivers/espi/espi_realtek_rts5912.c:
```diff
@@ -1240,6 +1464,80 @@ static void espi_vw_ch_setup(const struct device *dev)
+
+#define ESPI_VW_EVENT_RETRY_CNT   1024UL
+#define ESPI_VW_EVENT_TIMEOUT_CNT 10000UL
+
```
done

keith-zephyr commented:
@albertofloyd can you take another look at this?
